internal/admission/server.go

```diff
@@ -0,0 +1,179 @@
+	ValidateControlPlane(context.Context, operatorv1alpha1.ControlPlane) error
+	ValidateDataPlane(context.Context, operatorv1alpha1.DataPlane) error
+}
+
+// RequestHandler handles the requests of validating objects.
+type RequestHandler struct {
+	// Validator validates the entities that the k8s API-server asks
+	// it the server to validate.
+	Validator Validator
+	Logger    logr.Logger
+}
+
+// NewRequestHandler create a RequestHandler to handle validation requests.
+func NewRequestHandler(c client.Client, l logr.Logger) *RequestHandler {
+	return &RequestHandler{
+		Validator: &validator{
+			dataplaneValidator: dataplane.NewValidator(c),
+		},
+		Logger: l.WithValues("component", "validation-server"),
+	}
+}
+
+// ServeHTTP serves for HTTP requests.
+func (h *RequestHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Body == nil {
+		h.Logger.Error(fmt.Errorf("empty body"), "received request with empty body")
+		http.Error(w, "admission review object is missing", http.StatusBadRequest)
+		return
+	}
+	data, err := io.ReadAll(r.Body)
+	if err != nil {
+		h.Logger.Error(err, "failed to read request from client")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	review := &admissionv1.AdmissionReview{}
+	if err := json.Unmarshal(data, review); err != nil {
+		h.Logger.Error(err, "failed to parse AdmissionReview object")
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	response, err := h.handleValidation(r.Context(), review.Request)
+	if err != nil {
+		h.Logger.Error(err, "failed to run validation")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	review.Response = response
+	data, err = json.Marshal(review)
+	if err != nil {
+		h.Logger.Error(err, "failed to marshal response")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	_, err = w.Write(data)
+	if err != nil {
+		h.Logger.Error(err, "failed to write response")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+var (
+	controlPlaneGVResource = metav1.GroupVersionResource{
+		Group:    operatorv1alpha1.SchemeGroupVersion.Group,
+		Version:  operatorv1alpha1.SchemeGroupVersion.Version,
+		Resource: "controlplanes",
+	}
+	dataPlaneGVResource = metav1.GroupVersionResource{
+		Group:    operatorv1alpha1.SchemeGroupVersion.Group,
+		Version:  operatorv1alpha1.SchemeGroupVersion.Version,
+		Resource: "dataplanes",
+	}
+)
+
+func (h *RequestHandler) handleValidation(ctx context.Context, req *admissionv1.AdmissionRequest) (
+	*admissionv1.AdmissionResponse, error) {
+
+	deserializer := codecs.UniversalDeserializer()
+
+	var response admissionv1.AdmissionResponse
+	ok := true
+	msg := ""
+
+	if req == nil {
+		response.Allowed = false
+		response.Result = &metav1.Status{
+			Code:    http.StatusBadRequest,
+			Reason:  metav1.StatusReasonBadRequest,
+			Message: "empty request",
+			Status:  metav1.StatusFailure,
+		}
+		return &response, nil
+	}
```

Comment:
The `nil` check can be the very first thing done in this `func`.

```suggestion
	if req == nil {
		return &admissionv1.AdmissionResponse{
			Allowed: false
			Result: &metav1.Status{
				Code:    http.StatusBadRequest,
				Reason:  metav1.StatusReasonBadRequest,
				Message: "empty request",
				Status:  metav1.StatusFailure,
			}
		}, nil
	}
	
	var (
		response admissionv1.AdmissionResponse
		ok = true
		msg string
		deserializer = codecs.UniversalDeserializer()
	)
```

Reply:
updated.